### PR TITLE
Analytics events management

### PR DESCRIPTION
**Pull request by claudiulodro**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds a Google Analytics events management framework and wizard. We can refine it further, but this should be a good prototype right now. If the idea/architecture/ux looks good, we can refine it further.

*Disclaimer: Haven't actually tested much with a real site to verify analytics get recorded correctly. I'll do QA before switching this PR out of draft. The architecture has been my main concern so far.*

### How to test the changes in this Pull Request:

1. You should see a new Events tab in the Analytics wizard.

Screen Shot 2020-05-05 at 2 35 54 PM

I'm not a huge fan of those category labels, but that's what the NCS doc said to use for the names. Might be good to put something more readable for the event categories.

2. To simulate Site Kit AMP analytics, go to an AMP post and add the following snippet somewhere:

```

add_action( 'wp_footer', function(){
	$gtag_amp_opt = array(
		'vars'            => array(
			'gtag_id' => '123456',
			'config'  => array(
				'123456' => array(
					'groups' => 'default',
					'linker' => array(
						'domains' => array( 'https://newspack.test' ),
					),
				),
			),
		),
		'optoutElementId' => '__gaOptOutExtension',
	);

	$gtag_amp_opt_filtered = apply_filters( 'googlesitekit_amp_gtag_opt', $gtag_amp_opt );
	?>
	<amp-analytics type="gtag" data-credentials="include">
		<script type="application/json">
			<?php echo wp_json_encode( $gtag_amp_opt_filtered ); ?>
		</script>
	</amp-analytics>
	<?php

});

```

3. In the page source you should see something like:

```
<amp-analytics type="gtag" data-credentials="include">
   <script type="application/json">
   {"vars":{"gtag_id":"123456","config":{"123456":{"groups":"default","linker":{"domains":["https:\/\/newspack.test"]}}}},"optoutElementId":"__gaOptOutExtension","triggers":{"socialShareClickedFacebook":{"request":"event","on":"click","vars":{"event_name":"social share","event_label":"facebook","event_category":"NTG social"},"selector":"amp-social-share[type=\"facebook\"]"},"socialShareClickedTwitter":{"request":"event","on":"click","vars":{"event_name":"social share","event_label":"twitter","event_category":"NTG social"},"selector":"amp-social-share[type=\"twitter\"]"},"articleRead25":{"request":"event","on":"scroll","vars":{"event_name":"Vulputate sed per sapien eros urna venenatis turpis","event_label":"25%","event_category":"NTG article milestone"},"scrollSpec":{"verticalBoundaries":[25]}},"articleRead50":{"request":"event","on":"scroll","vars":{"event_name":"Vulputate sed per sapien eros urna venenatis turpis","event_label":"50%","event_category":"NTG article milestone"},"scrollSpec":{"verticalBoundaries":[50]}},"articleRead100":{"request":"event","on":"scroll","vars":{"event_name":"Vulputate sed per sapien eros urna venenatis turpis","event_label":"100%","event_category":"NTG article milestone"},"scrollSpec":{"verticalBoundaries":[100]}}}}		</script>
   </amp-analytics>
```

4. Remove that AMP-mode-specific code snippet from step 2. Go visit a non-AMP page.

5. In the page source you should see some scripts like like:

```
<script>
( function() {
    const elementSelector = 'a.share-twitter';
    const elements = document.querySelectorAll( elementSelector );
    
   for ( const element of elements ) {
       element.addEventListener( 'click', function() {
           ga( 'send', 'event', 'NTG social', 'social share', 'twitter' );
       } );
   }
} )();
</script>
 
etc.

```

4. If you deactivate an event category in the wizard, the events shouldn't show up in the page source on AMP and on non-AMP posts.

5. Other integration like popups can add events using [this filter](https://github.com/Automattic/newspack-plugin/compare/add/analytics?expand=1#diff-55af0c7c94b4a43e428b9ba731be0f2fR177), and they should show up automatically in the wizard and can be enabled/disabled there.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/class-analytics.php [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+class Analytics {
+
+	const EVENTS_ENABLED_OPTION          = 'newspack_analytics_events_enabled';
+	const EVENT_CATEGORY_DISABLED_OPTION = 'newspack_analytics_event_categories_disabled';
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		add_action( 'init', [ __CLASS__, 'register_events' ], 11 );
```
Comment (adekbadek) [inline includes/class-analytics.php:24]: There's no `register_events` method on this class – is something missing?

File: includes/class-analytics.php [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+				'event_category' => 'NTG article milestone',
+				'scrollSpec'     => [
+					'verticalBoundaries' => [ 50 ],
+				],
+			],
+			[
+				'id'             => 'articleRead100',
+				'on'             => 'scroll',
+				'event_name'     => get_the_title(),
+				'event_label'    => '100%',
```
Comment (adekbadek) [inline includes/class-analytics.php:166]: According to the [NTG Taxonomy](https://newsinitiative.withgoogle.com/training/states/ntg/assets/ntg-playbook.pdf#page=245): 
- the post title should be the `event_label`
- `n%` should be the `event_name` (assuming "Event Action" is `event_name`)
- there should be a `event_value` of the n percentage value (as integer)

**Comment by jeffersonrabb** [comment]
Here's a general thought. Maybe this branch should have no UI, only centralized management of the plumbing and the filters so that our other plugins can register GA events. In this model, consider this branch to be the "brain" with no visible UI. This will make it easier to review the PR, but more importantly avoids jumping to conclusions about where certain pieces of UI belong. For example, as we explore we may find that certain event tracking should be enabled in the theme, or the Pop-ups plugin. After we've actually implemented a few events, we'll probably have a clearer vision of which UI belongs where.

**Comment by claudiulodro** [comment]
Thanks for the feedback! I'll take the UI out and fix the rough edges @adekbadek noticed, then we can go from there.

**Comment by claudiulodro** [comment]
Closing in favor of #531 which fixes some issues and removes the UI.

**Pull request closed without merging**